x/deployment/types/events_test.go

```diff
@@ -0,0 +1,174 @@
+)
+
+type testEventParsing struct {
+	msg    sdkutil.Event
+	expErr error
+}
+
+func (tep testEventParsing) testMessageType() func(t *testing.T) {
+	ev, err := ParseEvent(tep.msg)
+	return func(t *testing.T) {
+		if tep.expErr != nil && err != tep.expErr {
+			if tep.expErr != errWildcard {
+				t.Errorf("unexpected error: %v exp: %v", err, tep.expErr)
+				t.Logf("%T %v", errors.Cause(err), err)
+			} else {
+				t.Logf("expected wildcard error: %v", err)
+			}
+		}
+		if ev == nil {
+			t.Logf("ParseEvent returned value is nil, TODO: return parsed data")
+		}
+	}
+}
+
+var TEPS = []testEventParsing{
+	{
+		msg: sdkutil.Event{
+			Type: "nil",
+		},
+		expErr: sdkutil.ErrUnknownType,
+	},
+	{
+		msg: sdkutil.Event{
+			Type: sdkutil.EventTypeMessage,
+		},
+		expErr: nil,
+	},
+
+	{
+		msg: sdkutil.Event{
+			Type:   sdkutil.EventTypeMessage,
+			Module: ModuleName,
+		},
+		expErr: nil,
+	},
+	{
+		msg: sdkutil.Event{
+			Type:   sdkutil.EventTypeMessage,
+			Module: "nil",
+		},
+		expErr: sdkutil.ErrUnknownModule,
+	},
+
+	{
+		msg: sdkutil.Event{
+			Type:   sdkutil.EventTypeMessage,
+			Module: ModuleName,
+			Action: "nil",
+		},
+		expErr: sdkutil.ErrUnknownAction,
+	},
+
+	{
+		msg: sdkutil.Event{
+			Type:   sdkutil.EventTypeMessage,
+			Module: ModuleName,
+			Action: evActionDeploymentCreate,
+			Attributes: []sdk.Attribute{
+				{
+					Key:   evOwnerKey,
+					Value: keyAcc.String(),
+				},
+				{
+					Key:   evDSeqKey,
+					Value: "5",
+				},
+			},
+		},
+		expErr: nil,
+	},
+	{
+		msg: sdkutil.Event{
+			Type:   sdkutil.EventTypeMessage,
+			Module: ModuleName,
+			Action: evActionDeploymentCreate,
+			Attributes: []sdk.Attribute{
+				{
+					Key:   evOwnerKey,
+					Value: keyAcc.String(),
+				},
+				{
+					Key:   evDSeqKey,
+					Value: "abc",
+				},
+			},
+		},
+		expErr: errWildcard,
+	},
+	{
+		msg: sdkutil.Event{
+			Type:   sdkutil.EventTypeMessage,
+			Module: ModuleName,
+			Action: evActionDeploymentCreate,
+			Attributes: []sdk.Attribute{
+				{
+					Key:   evOwnerKey,
+					Value: keyAcc.String(),
+				},
+			},
+		},
+		expErr: errWildcard,
+	},
+
+	{
+		msg: sdkutil.Event{
+			Type:   sdkutil.EventTypeMessage,
+			Module: ModuleName,
+			Action: evActionDeploymentUpdate,
+			Attributes: []sdk.Attribute{
+				{
+					Key:   evOwnerKey,
+					Value: keyAcc.String(),
+				},
+				{
+					Key:   evDSeqKey,
+					Value: "5",
+				},
+			},
+		},
+		expErr: nil,
+	},
+	{
+		msg: sdkutil.Event{
+			Type:   sdkutil.EventTypeMessage,
+			Module: ModuleName,
+			Action: evActionDeploymentUpdate,
+			Attributes: []sdk.Attribute{
+				{
+					Key:   evOwnerKey,
+					Value: keyAcc.String(),
+				},
+			},
+		},
+		expErr: errWildcard,
+	},
+}
+
+func TestEventParsing(t *testing.T) {
+	for i, test := range TEPS {
+		t.Run(fmt.Sprintf("%d", i), func(tep testEventParsing) func(t *testing.T) {
+			return func(t *testing.T) {
+				t.Run("parse-message-type", test.testMessageType())
```

Comment:
It's not clear to me why all these closures are needed?

```go
for i := range TEPS {
  t.Run("bla",func(t *testing.T){
      result, err := Parse(TEPS[i].msg)
      // ...
  })
}
```

you can also just add a message to the end of, say, `assert.Error(t,err,"blablabla")` - we use `github.com/stretchr/testify/assert`.

Reply:
Originally I was thinking if we wanted to have a step using the same test data, but different test label than `parse-message-type`, use the second layer of anonymous functions.
However this function being tested isn't worth additional separation. 

Removed.

Reply:
I believe this was the problem too @Ropes  - needed to use `TEPS[i]` instead of `test`